File: SurperPm-agent-web/backend/app/models/execution.py
"""Execution table model."""

import secrets
from datetime import UTC, datetime

from sqlalchemy import JSON, Column
from sqlmodel import Field, SQLModel


class Execution(SQLModel, table=True):
    __tablename__ = "execution"

    id: str = Field(default_factory=lambda: secrets.token_hex(8), primary_key=True)
    goal_id: int = Field(foreign_key="goal.id", nullable=False)
    workspace_id: str = Field(foreign_key="workspace.id", nullable=False)
    status: str = Field(default="pending")  # pending | running | success | failed | timeout
    branch: str | None = Field(default=None)
    started_at: datetime | None = Field(default=None)
    finished_at: datetime | None = Field(default=None)
    error: str | None = Field(default=None)
    log_path: str | None = Field(default=None)
    pr_url: str | None = Field(default=None)
    token_used: int | None = Field(default=None)
    token_budget: int | None = Field(default=None)
    summary: str | None = Field(default=None)
    artifacts: str | None = Field(default=None)
    logs: list | None = Field(default=None, sa_column=Column(JSON))
    created_at: datetime = Field(default_factory=lambda: datetime.now(UTC))

File: SurperPm-agent-web/frontend/src/lib/schemas/execution.ts
import { z } from "zod";

export const executionSchema = z.object({
  id: z.string(),
  goal_id: z.number(),
  workspace_id: z.string(),
  status: z.enum(["pending", "running", "success", "failed", "timeout"]),
  branch: z.string().nullable().optional(),
  started_at: z.string().nullable().optional(),
  finished_at: z.string().nullable().optional(),
  error: z.string().nullable().optional(),
  log_path: z.string().nullable().optional(),
  pr_url: z.string().nullable().optional(),
  token_used: z.number().nullable().optional(),
  token_budget: z.number().nullable().optional(),
  summary: z.string().nullable().optional(),
  artifacts: z.string().nullable().optional(),
  logs: z
    .array(
      z.object({
        kind: z.string(),
        text: z.string(),
        tool: z.string().optional(),
      })
    )
    .nullable()
    .optional(),
  created_at: z.string(),
});

export type Execution = z.infer<typeof executionSchema>;

export const executionListSchema = z.array(executionSchema);
